File: 322.零钱兑换.cpp
/*
 * @lc app=leetcode.cn id=322 lang=cpp
 *
 * [322] 零钱兑换
 */

// @lc code=start
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int minCoins[amount+1];
        minCoins[0] = 0;
        for (int i = 1; i <= amount; i++) {
            minCoins[i] = -1;
            for (int c : coins) {
                if (i - c >= 0 && minCoins[i-c] >= 0) {                    
                    minCoins[i] = minCoins[i] == -1 ? minCoins[i-c]+1 : min(minCoins[i], minCoins[i-c]+1);
                } 
            }
        }
        return minCoins[amount];
    }
};
// @lc code=end

File: 322.零钱兑换.py
#
# @lc app=leetcode.cn id=322 lang=python3
#
# [322] 零钱兑换
#

# @lc code=start
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        n = len(coins)
        dp = [0]*(amount+1)
        for i in range(1, amount+1):
            dp[i] = amount*2
            for j in range(n):
                if i-coins[j] >= 0 and dp[i-coins[j]] != -1:
                    dp[i] = min(dp[i], dp[i-coins[j]]+1)
            if dp[i] == amount*2:
                dp[i] = -1
        return dp[amount]
# @lc code=end

